src/lib.rs

```diff
@@ -281,52 +475,275 @@ pub fn get(
+#[cfg(all(feature = "client-reqwest", feature = "reqwest-blocking"))]
+pub struct DefaultRequestBuilder {
+    inner: RequestBuilder,
 }
 
-/// Send request to the server
-fn send(builder: RequestBuilder) -> Result<Vec<u8>> {
-    let mut response = builder.send()?;
-    if response.status() != StatusCode::OK {
-        return Err(Error::HttpStatus(response.status()));
+#[cfg(all(feature = "client-reqwest", feature = "reqwest-blocking"))]
+impl RequestBuildah for DefaultRequestBuilder {
+    type Error = Error;
+    type ReturnValue = String;
+    type ClientBuilder = ();
+    /// If the url is wrong then it will fail only during send
+    fn new(method: http::Method, url: &'_ str, _: &Self::ClientBuilder) -> Self {
+        let rb = CLIENT.request(method, Url::from_str(url).unwrap());
+        Self {
+            inner: rb
+        }
+    }
+
+    fn body(mut self, b: String) -> Self {
+        self.inner = self.inner.body(b);
+
+        self
+    }
+
+    fn header<K, V>(mut self, key: K, val: V) -> Self
+        where
+            HeaderName: TryFrom<K>,
+            HeaderValue: TryFrom<V>,
+            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
+            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>
+    {
+        self.inner = self.inner.header(key, val);
+
+        self
     }
-    let mut buf = vec![];
-    let _ = response.read_to_end(&mut buf)?;
-    Ok(buf)
+
+    fn send(self) -> std::result::Result<Self::ReturnValue, Error> {
+        let mut response = self.inner.send()?;
+        if response.status() != StatusCode::OK {
+            return Err(Error::HttpStatus(response.status()));
+        }
+        let mut buf = String::with_capacity(200);
+        let _ = response.read_to_string(&mut buf)?;
+        Ok(buf)
+    }
+}
+
+/// A generic request builder. Allows you to use any HTTP client.
+/// See [`DefaultRequestBuilder`] for one that uses [`reqwest::Client`].
+pub trait RequestBuildah {
+    /// The error produced while sending
+    type Error: Debug;
+
+    /// Generic return value allows you to return a future, allowing the possibility
+    /// of using this library in `async` environments.
+    type ReturnValue;
+
+    /// This is useful for reusing existing connection pools.
+    type ClientBuilder;
+
+    /// Construct the request builder
+    fn new(method: http::Method, url: &'_ str, client: &Self::ClientBuilder) -> Self;
+
+    /// Set the body
+    fn body(self, b: String) -> Self;
+
+    /// Set a header
+    fn header<K, V>(self, key: K, val: V) -> Self
+        where
+            HeaderName: TryFrom<K>,
+            HeaderValue: TryFrom<V>,
+            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
+            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>;
+
+    /// A `build`-like function that also sends the request
+    fn send(self) -> std::result::Result<Self::ReturnValue, Self::Error>;
+}
+
+
+#[macro_export]
+/// Counts number of var-args
+macro_rules! count {
```

Comment:
> Does this macro need to be a public API because it is used by the parse_query_string macro?

Yes.

> If so, please hide it from the documentation by adding the #[doc(hidden)] attribute.

I saw you remark about the possibility of making the `parse_query_string` macro a function. I'll try that, so maybe this can be removed too then.

Reply:
With https://github.com/gifnksm/oauth-client-rs/pull/44#discussion_r685102866 merged, `count!` macro can be removed like [this commit].

[this commit]: https://github.com/gifnksm/oauth-client-rs/commit/5c0782263a829d0ddf0ab1d158a92abd3e466d4c